blog: uuid (#1377)
questdb/questdb.io: blog/2023-02-10-uuid.mdx

---

> title: "UUID: Coordination-free unique keys"

Reviewer: I'd use book title case for first level title. And maybe this:

```suggestion
title: "UUID: Coordination-Free Unique Keys and Why They are Useful"
```

author: Jaromir Hamala
author_title: QuestDB Team
author_url: https://github.com/jerrinot
author_image_url: https://avatars.githubusercontent.com/jerrinot
description: Introduction to the UUID data type and coordination-free unique IDs
keywords:
  - UUID
  - distributed systems
  - coordination-free
  - keys
tags: [UUID, keys, coordination-free, distributed systems]
slug: uuid-coordination-free-unique-keys
---

In this article, [Jaromir Hamala](https://github.com/jerrinot) uses a real-world
scenario to highlight the use of the universally unique identifier (UUID). In
addition to an introduction to different versions of this data type, Jaromir
also showcases how using UUID in your QuestDB instance can help reduce query

> time and save storage space.

Reviewer: We discuss the need for an intro. This is only useful on the blog landing page.
This is a separate task relating to the layout of blog posts. I'd still keep some wording here for now as this is consistent and required to have the landing page function.

Author: agree

installation procedure. It would mean additional architectural complexity, but
it's doable. Sensor IDs are immutable, so each sensor needs to talk to the ID
service only once - right after the installation. That’s not too bad.

What if we need to store a unique ID for each data reading? Hitting the
centralized ID service whenever we need to store data is not an option. That
would stress the ID service too much and when the ID service is unavailable no
sensor could write any data.

What are the possible solutions?

In the simplest case, each sensor could talk to the remote ID service and
reserve a block of IDs it could then assign locally without further
coordination. When it exhausts the block, it asks the ID service for a new one.
This strategy would reduce the load on the ID service, and sensors could
function even when the ID service is temporarily unavailable. We could also
generate local reading IDs and prefix them with our unique immutable sensor ID.
We could also be smart and use fancy ID algorithms like FlakeIDs.

The strategies mentioned aim to minimize the need for coordination while still
making sure that the IDs are unique globally. The goal is to generate unique IDs

> without any coordination at all. This is what we call coordination-free unique

Reviewer: This makes me ask the question: Are all the UUID coordination free? The name 'UUID' does not really convey this piece of information. We could perhaps make this point clearer (either we are only focusing on the UUID that are coordination-free or all UUID are coordination free).

Author: I *think* all UUIDs can be considered coordination-free. 

Technically, the versions with a MAC address or wall-clock are not entirely coordination free: As MAC address assignment is coordinated, and so is time-tracking. but that's would be just nit-picking; I think they are coordination-free for practical purposes.

See version 7 for a UUID with a standard Unix epoch.

### Version 2

Version 2 is similar to version 1, but adds a local domain ID to the UUID. It's
not widely used.

### Versions 3 and 5

These versions use a hash function to generate the UUID. The hash function is
seeded with a namespace UUID and a name. The namespace UUID is used to ensure
uniqueness across multiple namespaces. The name is used to ensure uniqueness
within a namespace.

Version 3 uses MD5 as a hash function, while version 5 uses SHA-1. SHA-1
generates 160 bits so the digest is truncated to 128 bits.

### Version 4

Version 4 UUID is probably the most popular one. It relies solely on a random
generator to generate UUIDs, it's similar to the coin flip example above. This
means that the [quality of the random generator](https://xkcd.com/221/) is
critical.

### Version 6

Version 6 is similar to Version 1, but it has a different ordering of bytes. It
encodes the time from the most significant bits to the least significant bits.
This allows sorting UUIDs correctly by time when you sort just bytes
representing the UUIDs.

### Version 7

> Version 7 uses clock and random bits as inputs. Unlike versions 1, 2, or 6, it

Reviewer: I am not sure what clock means. Clock time?

Maybe (I googled it)

```suggestion
Version 7 uses a 48-bit timestamp and random data. Unlike versions 1, 2, or 6, it
```

uses a standard Unix epoch in milliseconds. It also uses a random generator
instead of a MAC address.

### Version 8

Version 8 is meant to be used for experimental and private use.

## Security considerations

UUIDs are designed to be unique, but they are not designed to be secret. What's
the difference? If you generate a UUID then you can assume it's different from
any other UUID generated before or after, but you should not treat them as a
password or a secret session identifier. This is what the
[RFC 4122](https://www.rfc-editor.org/rfc/rfc4122#page-16) says about this:

> Do not assume that UUIDs are hard to guess; they should not be used as
> security capabilities (identifiers whose mere possession grants access), for
> example. A predictable random number source will exacerbate the situation.

## UUID in QuestDB

> UUIDs are popular as synthetic IDs because they can be generated without any

Reviewer: ```suggestion
UUIDs are popular synthetic IDs because they can be generated without any
```

But I am not sure if the meaning may be changed...

Author: no, the meaning is still the same, thanks!

Storing UUIDs in a string column is possible, but it's inefficient. Let's do
some math: We already know each UUID has 128 bits, that's 16 bytes. The
canonical textual representation of UUID has 36 characters. QuestDB uses UTF-16
encoding for strings, so each ASCII character uses 2 bytes. There is also a
fixed cost of 4 bytes per string stored. So it takes 36 \* 2 + 4 = 76 bytes to
store a single UUID which contains just 16 bytes of information! It's not just
wasting disk space. QuestDB must read these bytes when evaluating a SQL
predicate, joining tables or calculating an aggregation. Thus storing UUIDs as
strings also makes your queries slower!

That's why QuestDB 6.7 implemented
[UUID](/docs/reference/sql/datatypes/#the-uuid-type) as a first-class data type.
This allows user applications to declare a column as UUID and then each UUID
stored will use only 16 bytes. Thanks to this, SQL queries will be faster.

## Demo time

Let’s create a table with a single string column and populate it with 1 billion
random UUIDs. The column is defined as the string type so the UUIDs will be
stored as strings:

```questdb-sql
CREATE TABLE tab_s (s string);

> INSERT INTO tab_s SELECT rnd_uuid4() FROM long_sequence(1000000000);

Reviewer: Does this work for you? I got timeout:

timeout, query aborted [fd=142]

Inserting less worked for me:
```suggestion
INSERT INTO tab_s SELECT rnd_uuid4() FROM long_sequence(100000000);
```

Author: hm, you might need to increase the query timeout. let me add a note about it.

Author: I pushed an update, let me know what you think.

---
title: "UUID: Coordination-Free Unique Keys and Why They are Useful"
author: Jaromir Hamala
author_title: QuestDB Team
author_url: https://github.com/jerrinot
author_image_url: https://avatars.githubusercontent.com/jerrinot
description: Introduction to the UUID data type and coordination-free unique IDs
keywords:
  - UUID
  - distributed systems
  - coordination-free
  - keys
tags: [UUID, keys, coordination-free, distributed systems]
slug: uuid-coordination-free-unique-keys
image: /img/blog/2023-02-10/banner.png
---

> Learn more about the universally unique identifier (UUID) and how using UUID in
> your QuestDB instance can help reduce query time and save storage space.

Reviewer: ```suggestion
Learn more about the universally unique identifier (UUID) and how using UUID in
your QuestDB instance can help reduce query time as well as save storage space.
```
